### PR TITLE
[prometheus] Fix issue with corrupted buffers when reading both OpenMetrics and plain text formats

### LINKED ISSUE #5515: Prometheus exported sometimes produces malformed data
# Bug Report

List of [all OpenTelemetry NuGet packages](https://www.nuget.org/profiles/OpenTelemetry) and version that you are using (e.g. `OpenTelemetry 1.0.2`):

```xml
    <PackageVersion Include="OpenTelemetry.Exporter.OpenTelemetryProtocol" Version="1.8.0" />
    <PackageVersion Include="OpenTelemetry.Exporter.Prometheus.AspNetCore" Version="1.8.0-rc.1" />
    <PackageVersion Include="OpenTelemetry.Extensions.Hosting" Version="1.8.0" />
    <PackageVersion Include="OpenTelemetry.Instrumentation.AspNetCore" Version="1.8.0" />
    <PackageVersion Include="OpenTelemetry.Instrumentation.EntityFrameworkCore" Version="1.0.0-beta.11" />
    <PackageVersion Include="OpenTelemetry.Instrumentation.Http" Version="1.8.0" />
    <PackageVersion Include="OpenTelemetry.Instrumentation.Process" Version="0.5.0-beta.5" />
    <PackageVersion Include="OpenTelemetry.Instrumentation.Runtime" Version="1.8.0" />
    <PackageVersion Include="OpenTelemetry.Instrumentation.StackExchangeRedis" Version="1.0.0-rc9.14" />
```

Runtime version: net8.0

## Symptom

Sometimes the exporter produces a malformed line:

```
dns_lookup_duration_seconds_bucket{otel_scope_name="System.Net.NameRe# TYPE otel_scope_info info
```

and causes Prometheus to fail: 

```
expected label value, got "\"System.Net.NameRe# TYPE otel_scope_info info\n" ("INVALID") while parsing: "dns_lookup_duration_seconds_bucket{otel_scope_name=\"System.Net.NameRe# TYPE otel_scope_info info\n"
```

**What is the expected behavior?**

There should not be malformed data in the metrics.

**What is the actual behavior?**

```
# TYPE dns_lookup_duration_seconds histogram
# UNIT dns_lookup_duration_seconds seconds
# HELP dns_lookup_duration_seconds Measures the time taken to perform a DNS lookup.
dns_lookup_duration_seconds_bucket{otel_scope_name="System.Net.NameRe# TYPE otel_scope_info info
# HELP otel_scope_info Scope metadata
otel_scope_info{otel_scope_name="System.Net.NameResolution"} 1
# TYPE otel_scope_info info
# HELP otel_scope_info Scope metadata
otel_scope_info{otel_scope_name="OpenTelemetry.Instrumentation.Runtime"} 1
# TYPE otel_scope_info info
# HELP otel_scope_info Scope metadata
otel_scope_info{otel_scope_name="OpenTelemetry.Instrumentation.Process"} 1
# TYPE otel_scope_info info
# HELP otel_scope_info Scope metadata
otel_scope_info{otel_scope_name="Microsoft.AspNetCore.Server.Kestrel"} 1
# TYPE otel_scope_info info
# HELP otel_scope_info Scope metadata
otel_scope_info{otel_scope_name="Microsoft.AspNetCore.Hosting"} 1
# TYPE otel_scope_info info
# HELP otel_scope_info Scope metadata
otel_scope_info{otel_scope_name="System.Net.Http"} 1
# TYPE otel_scope_info info
# HELP otel_scope_info Scope metadata
otel_scope_info{otel_scope_name="Microsoft.AspNetCore.RateLimiting"} 1
# TYPE otel_scope_info info
# HELP otel_scope_info Scope metadata
otel_scope_info{otel_scope_name="Microsoft.AspNetCore.Routing"} 1
# TYPE dns_lookup_duration_seconds histogram
# UNIT dns_lookup_duration_seconds seconds
# HELP dns_lookup_duration_seconds Measures the time taken to perform a DNS lookup.
dns_lookup_duration_seconds_bucket{otel_scope_name="System.Net.NameResolution",dns_question_name="localhost",le="0.005"} 10 1712549381.922
dns_lookup_duration_seconds_bucket{otel_scope_name="System.Net.NameResolution",dns_question_name="localhost",le="0.01"} 11 1712549381.922
dns_lookup_duration_seconds_bucket{otel_scope_name="System.Net.NameResolution",dns_question_name="localhost",le="0.025"} 11 1712549381.922
dns_lookup_duration_seconds_bucket{otel_scope_name="System.Net.NameResolution",dns_question_name="localhost",le="0.05"} 11 1712549381.922
dns_lookup_duration_seconds_bucket{otel_scope_name="System.Net.NameResolution",dns_question_name="localhost",le="0.075"} 11 1712549381.922
dns_lookup_duration_seconds_bucket{otel_scope_name="System.Net.NameResolution",dns_question_name="localhost",le="0.1"} 11 1712549381.922
dns_lookup_duration_seconds_bucket{otel_scope_name="System.Net.NameResolution",dns_question_name="localhost",le="0.25"} 11 1712549381.922
dns_lookup_duration_seconds_bucket{otel_scope_name="System.Net.NameResolution",dns_question_name="localhost",le="0.5"} 11 1712549381.922
dns_lookup_duration_seconds_bucket{otel_scope_name="System.Net.NameResolution",dns_question_name="localhost",le="0.75"} 11 1712549381.922
dns_lookup_duration_seconds_bucket{otel_scope_name="System.Net.NameResolution",dns_question_name="localhost",le="1"} 11 1712549381.922
dns_lookup_duration_seconds_bucket{otel_scope_name="System.Net.NameResolution",dns_question_name="localhost",le="2.5"} 11 1712549381.922
dns_lookup_duration_seconds_bucket{otel_scope_name="System.Net.NameResolution",dns_question_name="localhost",le="5"} 11 1712549381.922
dns_lookup_duration_seconds_bucket{otel_scope_name="System.Net.NameResolution",dns_question_name="localhost",le="7.5"} 11 1712549381.922
dns_lookup_duration_seconds_bucket{otel_scope_name="System.Net.NameResolution",dns_question_name="localhost",le="10"} 11 1712549381.922
dns_lookup_duration_seconds_bucket{otel_scope_name="System.Net.NameResolution",dns_question_name="localhost",le="+Inf"} 11 1712549381.922
dns_lookup_duration_seconds_sum{otel_scope_name="System.Net.NameResolution",dns_question_name="localhost"} 0.018864899999999997 1712549381.922
dns_lookup_duration_seconds_count{otel_scope_name="System.Net.NameResolution",dns_question_name="localhost"} 11 1712549381.922
```

## Reproduce

Simply adding some instrumentations.

```cs
            var otel = services.AddOpenTelemetry();

            otel.ConfigureResource(resource => resource.AddService("foo"));

            otel.WithMetrics(metrics =>
            {
                metrics.AddAspNetCoreInstrumentation();
                metrics.AddHttpClientInstrumentation();
                metrics.AddRuntimeInstrumentation();
                metrics.AddProcessInstrumentation();

                metrics.AddPrometheusExporter(options =>
                {
                    options.DisableTotalNameSuffixForCounters = true;
                });
            });

            otel.WithTracing(tracing =>
            {
                tracing.AddAspNetCoreInstrumentation();
                tracing.AddHttpClientInstrumentation();
                tracing.AddEntityFrameworkCoreInstrumentation();
                tracing.AddRedisInstrumentation();
            });

            otel.UseOtlpExporter(...);

            // ...
            // var app = builder.Build();

            app.UseOpenTelemetryPrometheusScrapingEndpoint();
```

### PR DESCRIPTION
**Pull request by robertcoltheart**
Fixes #5515 

## Changes

Separate prometheus buffers into plain text and openmetrics. This is the first part of splitting #5517. It also optimizes the generation of the buffers, and only generates the requested buffer, rather than generating both simultaneously. 

## Merge requirement checklist

* [X] [CONTRIBUTING](https://github.com/open-telemetry/opentelemetry-dotnet/blob/main/CONTRIBUTING.md) guidelines followed (license requirements, nullable enabled, static analysis, etc.)
* [X] Unit tests added/updated
* [X] Appropriate `CHANGELOG.md` files updated for non-trivial changes

### CODE REVIEW

**Review by reyang — APPROVED** [review]

**Comment by reyang** [comment]
Thanks @robertcoltheart!

File: src/OpenTelemetry.Exporter.Prometheus.HttpListener/CHANGELOG.md [hunk 1]
```diff
@@ -2,6 +2,10 @@
 
 ## Unreleased
 
+* Fixed an issue with corrupted buffers when reading both OpenMetrics and
+  plain text formats formats from Prometheus exporters
```
Comment (Kielek) [inline src/OpenTelemetry.Exporter.Prometheus.HttpListener/CHANGELOG.md:6]:
```suggestion
* Fixed an issue with corrupted buffers when reading both OpenMetrics and
  plain text formats from Prometheus exporters.
```

File: test/OpenTelemetry.Exporter.Prometheus.AspNetCore.Tests/PrometheusExporterMiddlewareTests.cs [hunk 2]
```diff
@@ -248,6 +248,46 @@ public Task PrometheusExporterMiddlewareIntegration_UseOpenMetricsVersionHeader(
...
+
+        using var meter = new Meter(MeterName, MeterVersion);
+
+        var beginTimestamp = DateTimeOffset.Now.ToUnixTimeMilliseconds();
+
+        var counter = meter.CreateCounter<double>("counter_double");
+        counter.Add(100.18D, tags);
+        counter.Add(0.99D, tags);
+
+        var testCases = new bool[] { true, false };
```
Comment (vishweshbankwar) [inline test/OpenTelemetry.Exporter.Prometheus.AspNetCore.Tests/PrometheusExporterMiddlewareTests.cs:271]: add more values here. {true, false, true, true, false}. 

The issue was surfacing due to this sequence > OpenMetrics, Plaintext, OpenMetrics (Plaintext was overwriting OpenMetrics data (specifically scope information).

**Comment by robertcoltheart** [comment]
Done, please review

**Review by vishweshbankwar — APPROVED** [review]
LGTM - Thanks @robertcoltheart!

**Review by Kielek — APPROVED** [review]

**Comment by bbrandt** [comment]
([orig](https://github.com/open-telemetry/opentelemetry-dotnet/commit/8177a391a662f08352f40e8c4552dc1eb07a8cc4#r143411365))
This change results in a regression causing no output to be displayed when IncreaseBufferSize(...) is needed. The issue seems to be that IncreaseBufferSize(...) is taking a reference to the buffer local variable, but not to this.openMetricsBuffer or this.plainTextBuffer, so these members do not get the updated content after a resize occurs. Then [on 294](https://github.com/open-telemetry/opentelemetry-dotnet/commit/8177a391a662f08352f40e8c4552dc1eb07a8cc4#diff-d96b89d4588c0c37b9a6155b5eb2403cc3d5e475028cdb49749d345840508a65R294) when we try to access this.openMetricsBuffer or this.plainTextBuffer it is the original buffer instead of the resized buffer.

Looks like this is fixed by #5676, which has not yet been merged.

@Kielek @vishweshbankwar @CodeBlanch What is needed to get #5676 merged?

**Pull request merged**